[LGWebOS] Firmware 03.30.16 requires use of secured web socket

After updating one of my TVs to the latest LG software the binding now shows OFFLINE
My other LG TV, which was not updated, continues to work as expected.
The TV still responds to the wake-on-LAN function.

debug log from the binding
`[DEBUG] [webos.internal.handler.LGWebOSHandler] - Connection failed - error: HttpConnectionOverHTTP@c02c3e::SocketChannelEndPoint@ff6946{l=/192.168.1.108:34790,r=/192.168.1.136:3000,ISHUT,fill=-,flush=W,to=0/0}{io=0/0,kio=0,kro=1}->HttpConnectionOverHTTP@c02c3e(l:/192.168.1.108:34790 <-> r:/192.168.1.136:3000,closed=false)=>HttpChannelOverHTTP@15d50eb(exchange=HttpExchange@1e17637{req=WebSocketUpgradeRequest[GET / HTTP/1.1]@691de9[TERMINATED/null] res=HttpResponse[null 0 null]@1e7e2fb[PENDING/null]})[send=HttpSenderOverHTTP@2a2464(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@c0ac18{s=START}],recv=HttpReceiverOverHTTP@967e7c(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]`

Numerous other users are reporting the same symptoms .

Looking at the Home Assistant forum, users reporting a similar issue.

Apparently the issue may be caused by LG now requiring a secure web socket
https://github.com/home-assistant-libs/aiowebostv/pull/112

Just changing to port 3001 and `wss://` leads to the actual problem that needs to be solved regarding the self-signed certificate:

```
2023-02-02 22:04:11.493 [DEBUG] [bos.internal.handler.LGWebOSTVSocket] - Connecting to: wss://lgtv:3001
2023-02-02 22:04:11.555 [TRACE] [bos.internal.handler.LGWebOSTVSocket] - Connection Error
javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.ssl.Alert.createSSLException(Alert.java:131) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:353) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:296) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:291) ~[?:?]
	at sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:654) ~[?:?]
	at sun.security.ssl.CertificateMessage$T12CertificateConsumer.onCertificate(CertificateMessage.java:473) ~[?:?]
	at sun.security.ssl.CertificateMessage$T12CertificateConsumer.consume(CertificateMessage.java:369) ~[?:?]
	at sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:392) ~[?:?]
	at sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:443) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1076) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1063) ~[?:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:1010) ~[?:?]
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.fill(SslConnection.java:654) ~[?:?]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:168) ~[?:?]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:80) ~[?:?]
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:131) ~[?:?]
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:172) ~[?:?]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) ~[?:?]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) ~[?:?]
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:555) ~[?:?]
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:410) ~[?:?]
	at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:164) ~[?:?]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) ~[?:?]
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104) ~[?:?]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338) ~[bundleFile:9.4.46.v20220331]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315) ~[bundleFile:9.4.46.v20220331]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173) ~[bundleFile:9.4.46.v20220331]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131) ~[bundleFile:9.4.46.v20220331]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409) ~[bundleFile:9.4.46.v20220331]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883) [bundleFile:9.4.46.v20220331]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034) [bundleFile:9.4.46.v20220331]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:439) ~[?:?]
	at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:306) ~[?:?]
	at sun.security.validator.Validator.validate(Validator.java:264) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:313) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:276) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:141) ~[?:?]
	at org.openhab.core.io.net.http.internal.ExtensibleTrustManagerImpl.checkServerTrusted(ExtensibleTrustManagerImpl.java:123) ~[?:?]
	at sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:632) ~[?:?]
	... 28 more
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:141) ~[?:?]
	at sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:126) ~[?:?]
	at java.security.cert.CertPathBuilder.build(CertPathBuilder.java:297) ~[?:?]
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:434) ~[?:?]
	at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:306) ~[?:?]
	at sun.security.validator.Validator.validate(Validator.java:264) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:313) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:276) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:141) ~[?:?]
	at org.openhab.core.io.net.http.internal.ExtensibleTrustManagerImpl.checkServerTrusted(ExtensibleTrustManagerImpl.java:123) ~[?:?]
	at sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:632) ~[?:?]
	... 28 more
2023-02-02 22:04:11.607 [DEBUG] [ebos.internal.handler.LGWebOSHandler] - Connection failed - error: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
```

--- a/bundles/org.openhab.binding.lgwebos/README.md
+++ b/bundles/org.openhab.binding.lgwebos/README.md
@@ -36,11 +36,12 @@ WebOS TV has three configuration parameters.
 
 Parameters:
 
-| Name       | Description                                                                                         |
-|------------|-----------------------------------------------------------------------------------------------------|
-| host       | Hostname or IP address of TV                                                                        |
-| key        | Key exchanged with TV after pairing (enter it after you paired the device)                          |
-| macAddress | The MAC address of your TV to turn on via Wake On Lan (WOL). The binding will attempt to detect it. |
+| Name       | Description                                                                                                                                                                    |
+|------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| host       | Hostname or IP address of TV                                                                                                                                                   |
+| key        | Key exchanged with TV after pairing (enter it after you paired the device)                                                                                                     |
+| macAddress | The MAC address of your TV to turn on via Wake On Lan (WOL). The binding will attempt to detect it.                                                                            |
+| useTLS     | Enable Transport Layer Security. This is required by latest firmware versions and should work with older versions as well. In case of compatibility issues it can be disabled. |
 
 ### Configuration in .things file
 

--- a/bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/LGWebOSBindingConstants.java
+++ b/bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/LGWebOSBindingConstants.java
@@ -68,4 +68,7 @@ public class LGWebOSBindingConstants {
     public static final String CHANNEL_MEDIA_STOP = "mediaStop";
     public static final String CHANNEL_APP_LAUNCHER = "appLauncher";
     public static final String CHANNEL_RCBUTTON = "rcButton";
+
+    public static final int DEFAULT_WS_PORT = 3000;
+    public static final int DEFAULT_WSS_PORT = 3001;
 }

--- a/bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/LGWebOSHandlerFactory.java
+++ b/bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/LGWebOSHandlerFactory.java
@@ -16,6 +16,8 @@ import static org.openhab.binding.lgwebos.internal.LGWebOSBindingConstants.*;
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.util.ssl.SslContextFactory;
 import org.eclipse.jetty.websocket.client.WebSocketClient;
 import org.openhab.binding.lgwebos.internal.handler.LGWebOSHandler;
 import org.openhab.core.io.net.http.WebSocketFactory;
@@ -53,7 +55,8 @@ public class LGWebOSHandlerFactory extends BaseThingHandlerFactory {
          * Cannot use openHAB's shared web socket client (webSocketFactory.getCommonWebSocketClient()) as we have to
          * change client settings.
          */
-        this.webSocketClient = webSocketFactory.createWebSocketClient("lgwebos");
+        var httpClient = new HttpClient(new SslContextFactory.Client(true));
+        this.webSocketClient = new WebSocketClient(httpClient);
         this.stateDescriptionProvider = stateDescriptionProvider;
     }
 
@@ -74,9 +77,6 @@ public class LGWebOSHandlerFactory extends BaseThingHandlerFactory {
     @Override
     protected void activate(ComponentContext componentContext) {
         super.activate(componentContext);
-        // LGWebOS TVs only support WEAK cipher suites, thus not using SSL.
-        // SslContextFactory sslContextFactory = new SslContextFactory(true);
-        // sslContextFactory.addExcludeProtocols("tls/1.3");
 
         // reduce timeout from default 15sec
         this.webSocketClient.setConnectTimeout(1000);

--- a/bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSConfiguration.java
+++ b/bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSConfiguration.java
@@ -25,11 +25,11 @@ import org.eclipse.jdt.annotation.Nullable;
 public class LGWebOSConfiguration {
     @Nullable
     String host; // name has to match LGWebOSBindingConstants.CONFIG_HOST
-    int port = 3000; // 3001 for TLS
     @Nullable
     String key; // name has to match LGWebOSBindingConstants.CONFIG_KEY
     @Nullable
     String macAddress; // name has to match LGWebOSBindingConstants.CONFIG_MAC_ADDRESS
+    boolean useTLS = true;
 
     public String getHost() {
         String h = host;
@@ -41,8 +41,8 @@ public class LGWebOSConfiguration {
         return k == null ? "" : k;
     }
 
-    public int getPort() {
-        return port;
+    public boolean getUseTLS() {
+        return useTLS;
     }
 
     public String getMacAddress() {
@@ -52,7 +52,7 @@ public class LGWebOSConfiguration {
 
     @Override
     public String toString() {
-        return "WebOSConfiguration [host=" + host + ", port=" + port + ", key.length=" + getKey().length()
+        return "WebOSConfiguration [host=" + host + ", useTLS=" + useTLS + ", key.length=" + getKey().length()
                 + ", macAddress=" + macAddress + "]";
     }
 }

--- a/bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java
+++ b/bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java
@@ -135,7 +135,7 @@ public class LGWebOSHandler extends BaseThingHandler
             return;
         }
 
-        LGWebOSTVSocket s = new LGWebOSTVSocket(webSocketClient, this, host, c.getPort(), scheduler);
+        LGWebOSTVSocket s = new LGWebOSTVSocket(webSocketClient, this, host, c.getUseTLS(), scheduler);
         s.setListener(this);
         socket = s;
 

--- a/bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java
+++ b/bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java
@@ -116,6 +116,7 @@ public class LGWebOSTVSocket {
     private final URI destUri;
     private final LGWebOSTVKeyboardInput keyboardInput;
     private final ScheduledExecutorService scheduler;
+    private final Protocol protocol;
 
     public enum State {
         DISCONNECTING,
@@ -123,6 +124,19 @@ public class LGWebOSTVSocket {
         CONNECTING,
         REGISTERING,
         REGISTERED
+    }
+
+    private enum Protocol {
+        WEB_SOCKET("ws", DEFAULT_WS_PORT),
+        WEB_SOCKET_SECURE("wss", DEFAULT_WSS_PORT);
+
+        private Protocol(String name, int port) {
+            this.name = name;
+            this.port = port;
+        }
+
+        public String name;
+        public int port;
     }
 
     private State state = State.DISCONNECTED;
@@ -140,14 +154,15 @@ public class LGWebOSTVSocket {
 
     private @Nullable ScheduledFuture<?> disconnectingJob;
 
-    public LGWebOSTVSocket(WebSocketClient client, ConfigProvider config, String host, int port,
+    public LGWebOSTVSocket(WebSocketClient client, ConfigProvider config, String host, boolean useTLS,
             ScheduledExecutorService scheduler) {
         this.config = config;
         this.client = client;
         this.keyboardInput = new LGWebOSTVKeyboardInput(this);
+        this.protocol = useTLS ? Protocol.WEB_SOCKET_SECURE : Protocol.WEB_SOCKET;
 
         try {
-            this.destUri = new URI("ws://" + host + ":" + port);
+            this.destUri = new URI(protocol.name + "://" + host + ":" + protocol.port);
         } catch (URISyntaxException e) {
             throw new IllegalArgumentException("IP address or hostname provided is invalid: " + host);
         }
@@ -916,8 +931,13 @@ public class LGWebOSTVSocket {
             @Override
             public void onSuccess(@Nullable JsonObject jsonObj) {
                 if (jsonObj != null) {
-                    String socketPath = jsonObj.get("socketPath").getAsString().replace("wss:", "ws:").replace(":3001/",
-                            ":3000/");
+                    String socketPath = jsonObj.get("socketPath").getAsString();
+                    if (protocol == Protocol.WEB_SOCKET) {
+                        socketPath = socketPath
+                                .replace(Protocol.WEB_SOCKET_SECURE.name + ":", Protocol.WEB_SOCKET.name + ":")
+                                .replace(":" + Protocol.WEB_SOCKET_SECURE.port + "/",
+                                        ":" + Protocol.WEB_SOCKET.port + "/");
+                    }
                     try {
                         mouseSocket.connect(new URI(socketPath));
                     } catch (URISyntaxException e) {

--- a/bundles/org.openhab.binding.lgwebos/src/main/resources/OH-INF/config/config.xml
+++ b/bundles/org.openhab.binding.lgwebos/src/main/resources/OH-INF/config/config.xml
@@ -22,6 +22,13 @@
 				(WOL) when it receives command ON on channel power. Accepted value is six groups of two hexadecimal digits,
 				separated by hyphens or colons, e.g '3c:cd:93:c2:20:e0'.</description>
 		</parameter>
+		<parameter name="useTLS" type="boolean" required="false">
+			<label>Use TLS</label>
+			<description>Enable Transport Layer Security. This is required by latest firmware versions and should work with older
+				versions as well. In case of compatibility issues it can be disabled.</description>
+			<advanced>true</advanced>
+			<default>true</default>
+		</parameter>
 	</config-description>
 
 </config-description:config-descriptions>

--- a/bundles/org.openhab.binding.lgwebos/src/main/resources/OH-INF/i18n/lgwebos.properties
+++ b/bundles/org.openhab.binding.lgwebos/src/main/resources/OH-INF/i18n/lgwebos.properties
@@ -16,6 +16,8 @@ thing-type.config.lgwebos.WebOSTV.key.label = Access Key
 thing-type.config.lgwebos.WebOSTV.key.description = Key exchanged with TV after pairing.
 thing-type.config.lgwebos.WebOSTV.macAddress.label = MAC Address
 thing-type.config.lgwebos.WebOSTV.macAddress.description = If MAC Address of TV is entered here, the binding will attempt to power on the device via Wake On Lan (WOL) when it receives command ON on channel power. Accepted value is six groups of two hexadecimal digits, separated by hyphens or colons, e.g '3c:cd:93:c2:20:e0'.
+thing-type.config.lgwebos.WebOSTV.useTLS.label = Use TLS
+thing-type.config.lgwebos.WebOSTV.useTLS.description = Enable Transport Layer Security. This is required by latest firmware versions and should work with older versions as well. In case of compatibility issues it can be disabled.
 
 # channel types
 
@@ -69,7 +71,7 @@ actionShowToastWithIconDesc = Sends a toast message to a WebOS device with custo
 actionShowToastInputIconLabel = Icon
 actionShowToastInputIconDesc = The URL to the icon to display
 
-# Thing status descriptions
+# thing status descriptions
 
 offline.config-error-unknown-host = Missing parameter "host"
 offline.comm-error-connexion-failed = Connection Failed: {0}

